# Exercise sheet 4
**Hello everyone!**

**Points: 15**

Topics of this exercise sheet are:
* Working with seaborn
* Data visualization principles

Please let us know if you have questions or problems! <br>
Contact us during the exercise session, on [ADAM](https://adam.unibas.ch/goto_adam_crs_1266890.html), [Piazza](https://piazza.com/class/kzy15kp8s5t6ku), or [via email](https://sada.dmi.unibas.ch/en/teaching/pids22).

Please submit this exercise sheet on **ADAM**.
Naming conventions:
* Name of notebook: "Exercise sheet 4.ipynb"
* Please put it in a folder structure like this: "your_unibas_short_name/Exercise4/Exercise sheet 4.ipynb"
* You can find your unibas short name either...
    * by looking at the filename of the feedback on exercise sheet 2 (from ADAM)
    * or if you didn't submit exercise sheet 2, log in to viaweb.unibas.ch and you will see your unibas short name after the welcome message
* Then zip this folder and upload it on ADAM.

**Handout date**: 2022/03/28 <br>
**Submission date**: 2022/04/04 <br>
**Discussion date**: 2022/04/05

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
from nose.tools import assert_is_instance, assert_equal, assert_almost_equal, assert_true

## 1 (5 points)

### 1a (1 point)
We will use data from [Kaggle](https://www.kaggle.com/datasets/prasertk/healthy-lifestyle-cities-report-2021) about different cities around the world.

In [2]:
cities = pd.read_csv("healthy_lifestyle_city_2021.csv")
cities.head()

City  Rank Sunshine hours(City) Cost of a bottle of water(City)  \
0   Amsterdam     1                 1858                           £1.92   
1      Sydney     2                 2636                           £1.48   
2      Vienna     3                 1884                           £1.94   
3   Stockholm     4                 1821                           £1.72   
4  Copenhagen     5                 1630                           £2.19   

  Obesity levels(Country)  Life expectancy(years) (Country)  \
0                  20.40%                              81.2   
1                  29.00%                              82.1   
2                  20.10%                              81.0   
3                  20.60%                              81.8   
4                  19.70%                              79.8   

  Pollution(Index score) (City) Annual avg. hours worked  \
0                         30.93                     1434   
1                         26.86                     1712   
2                         17.33                     1501   
3                         19.63                     1452   
4                         21.24                     1380   

   Happiness levels(Country)  Outdoor activities(City)  \
0                       7.44                       422   
1                       7.22                       406   
2                       7.29                       132   
3                       7.35                       129   
4                       7.64                       154   

   Number of take out places(City) Cost of a monthly gym membership(City)  
0                             1048                                 £34.90  
1                             1103                                 £41.66  
2                             1008                                 £25.74  
3                              598                                 £37.31  
4                              523                                 £32.53

First, let's clean this data a little.
* Replace the values in "Cost of a bottle of water(City)", "Sunshine hours(City)" and "Cost of a monthly gym membership(City)" with numerical values (hint: use the pandas functions .str and .as_type())
* Replace the values in "Obesity levels(Country)" to be numerical values. A value "20.40%" should be converted to "0.204".

In [3]:
# YOUR CODE HERE
raise NotImplementedError()

NotImplementedError: 

In [ ]:
assert_is_instance(cities, pd.DataFrame)
assert_equal(len(cities), 44)
assert_almost_equal(cities.loc[(cities["City"] == "Vienna")]["Sunshine hours(City)"].values[0], 1884.0)
assert_almost_equal(cities.loc[(cities["City"] == "Stockholm")]["Cost of a bottle of water(City)"].values[0], 1.72)
assert_almost_equal(cities.loc[(cities["City"] == "Milan")]["Obesity levels(Country)"].values[0], 0.199)
assert_almost_equal(cities.loc[(cities["City"] == "New York")]["Cost of a monthly gym membership(City)"].values[0], 64.66)

### 1b (1 point)
What is the mean of the column "Obesity levels(Country)"?

In [ ]:
# mean_obesity_level = ...

# YOUR CODE HERE
raise NotImplementedError()

### 1c (1 point)
Is the mean a good measure to summarize the values of "Obesity levels(Country)"?

In [ ]:
# a visualization might help...

# YOUR CODE HERE
raise NotImplementedError()

YOUR ANSWER HERE

### 1d (2 points)
Please make a new column called "Continent" containing the continent of each city. <br>
If a city is in multiple continents, choose one (it will make the following plots cleaner). <br>
Use this list of continent names: Europe, Australia, America, Asia, Antarctica, Oceania.

Hint: You can go two routes, find a different dataframe on the internet to join to this one (using the pandas function merge), or hand-code the information. Whichever is easier for you :)

In [ ]:
# cities["Continent"] = ...

# YOUR CODE HERE
raise NotImplementedError()

In [ ]:
assert_equal(cities.loc[(cities["City"] == "Vienna")]["Continent"].values[0], "Europe")
assert_equal(cities.loc[(cities["City"] == "Sydney")]["Continent"].values[0], "Australia")
assert_equal(cities.loc[(cities["City"] == "Johannesburg")]["Continent"].values[0], "Africa")
assert_equal(cities.loc[(cities["City"] == "Sao Paulo")]["Continent"].values[0], "America")
assert_equal(cities.loc[(cities["City"] == "Shanghai")]["Continent"].values[0], "Asia")

What does the column "Continent" have to do with question 1c? (Hint: This is pretty similar to the penguin flipper lengths from the lecture)

In [ ]:
# a visualization might help...

# YOUR CODE HERE
raise NotImplementedError()

YOUR ANSWER HERE

![](https://upload.wikimedia.org/wikipedia/commons/8/89/Melbourne_skyline_from_Williamstown.png)

## 2 (8 points)
You have gotten the job as data visualization expert in Melbourne. 🎉 Congratulations! <br>
Your first task is to make a stunning plot to convince everyone that Melbourne is the best city in the whole wide world! We want to print it on posters! Or at least in the national tourism brochure.

* Make **one plot** using any and all (evil) principles of data visualization. If you need inspiration, look at the examples on the lecture slides or here: https://www.youtube.com/watch?v=E91bGT9BjYk or google "misleading data visualization".
* If necessary, write (max) 3 sentences to explain your plot

In [ ]:
# Put your plot here

# YOUR CODE HERE
raise NotImplementedError()

Describe your plot here

## 3 (2 points)
### 3a (2 points)
Where did this data even come from? Find one instance where a data point in the cities dataframe is false. Write one sentence explaining why it is false and provide a source proving your claim.

YOUR ANSWER HERE